src/core/Entry.cpp

```diff
@@ -736,66 +834,81 @@ QString Entry::maskPasswordPlaceholders(const QString &str) const
-            result.replace(result,attributes()->value(key));
-            break;
-        }
-    }
-
-    // resolving references in format: {REF:<WantedField>@I:<uuid of referenced entry>}
-    // using format from http://keepass.info/help/base/fieldrefs.html at the time of writing,
-    // but supporting lookups of standard fields and references by UUID only
-
-    QRegExp* tmpRegExp = m_attributes->referenceRegExp();
-    if (tmpRegExp->indexIn(result) != -1) {
-        // cap(0) contains the whole reference
-        // cap(1) contains which field is wanted
-        // cap(2) contains the uuid of the referenced entry
-        Entry* tmpRefEntry = m_group->database()->resolveEntry(Uuid(QByteArray::fromHex(tmpRegExp->cap(2).toLatin1())));
-        if (tmpRefEntry) {
-            // entry found, get the relevant field
-            QString tmpRefField = tmpRegExp->cap(1).toLower();
-            if (tmpRefField == "t") result.replace(tmpRegExp->cap(0), tmpRefEntry->title(), Qt::CaseInsensitive);
-            else if (tmpRefField == "u") result.replace(tmpRegExp->cap(0), tmpRefEntry->username(), Qt::CaseInsensitive);
-            else if (tmpRefField == "p") result.replace(tmpRegExp->cap(0), tmpRefEntry->password(), Qt::CaseInsensitive);
-            else if (tmpRefField == "a") result.replace(tmpRegExp->cap(0), tmpRefEntry->url(), Qt::CaseInsensitive);
-            else if (tmpRefField == "n") result.replace(tmpRegExp->cap(0), tmpRefEntry->notes(), Qt::CaseInsensitive);
-        }
-    }
-
+    return resolveMultiplePlaceholdersRecursive(str, ResolveMaximumDepth);
+}
+
+QString Entry::resolvePlaceholder(const QString& placeholder) const
+{
+    return resolvePlaceholderRecursive(placeholder, ResolveMaximumDepth);
+}
+
+QString Entry::resolveUrlPlaceholder(const QString &str, Entry::PlaceholderType placeholderType) const
+{
+    if (str.isEmpty())
+        return QString();
+
+    const QUrl qurl(str);
+    switch (placeholderType) {
+    case PlaceholderType::UrlWithoutScheme:
+        return qurl.toString(QUrl::RemoveScheme | QUrl::FullyDecoded);
+    case PlaceholderType::UrlScheme:
+        return qurl.scheme();
+    case PlaceholderType::UrlHost:
+        return qurl.host();
+    case PlaceholderType::UrlPort:
+        return QString::number(qurl.port());
+    case PlaceholderType::UrlPath:
+        return qurl.path();
+    case PlaceholderType::UrlQuery:
+        return qurl.query();
+    case PlaceholderType::UrlFragment:
+        return qurl.fragment();
+    case PlaceholderType::UrlUserInfo:
+        return qurl.userInfo();
+    case PlaceholderType::UrlUserName:
+        return qurl.userName();
+    case PlaceholderType::UrlPassword:
+        return qurl.password();
+    default:
+        Q_ASSERT(false);
+        break;
+    }
+
+    return QString();
+}
+
+Entry::PlaceholderType Entry::placeholderType(const QString &placeholder) const
+{
+    if (!placeholder.startsWith(QLatin1Char('{')) || !placeholder.endsWith(QLatin1Char('}'))) {
```

Comment:
Can this case ever occur?

Reply:
yes. In some places, the function "resolvePlaceholder" called directly.